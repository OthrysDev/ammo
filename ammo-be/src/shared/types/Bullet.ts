// /!\ We cannot extend ConnectorRequest here, as this file is exported to the front-end, whereas ConnectorRequest is not

export type Bullet = {
    id: string;
    date: Date;
    url: string;
    method:
        | 'GET'
        | 'PUT'
        | 'POST'
        | 'DELETE'
        | 'HEAD'
        | 'CONNECT'
        | 'OPTIONS'
        | 'TRACE'
        | 'PATCH';
    request: {
        headers: Record<string, string | string[]>;
        body?: unknown;
    };
    response: {
        headers: Record<string, string | string[]>;
        body?: unknown;
        status?: number;
    };
};
