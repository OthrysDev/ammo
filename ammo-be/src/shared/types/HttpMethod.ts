type HttpMethod =
    | 'GET'
    | 'PUT'
    | 'POST'
    | 'DELETE'
    | 'HEAD'
    | 'CONNECT'
    | 'OPTIONS'
    | 'TRACE'
    | 'PATCH';

export default HttpMethod;
