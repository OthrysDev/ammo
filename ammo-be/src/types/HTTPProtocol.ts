type HTTPProtocol = 'http' | 'https';

export default HTTPProtocol;
