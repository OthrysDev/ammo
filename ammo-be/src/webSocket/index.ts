import io from 'socket.io';
import http from 'http';

let ioServer: io.Server;

const initWS = (server: http.Server): void => {
    ioServer = new io.Server(server, {
        cors: {
            origin: ['http://localhost:3000'],
        },
    });
};

export { initWS, ioServer };
