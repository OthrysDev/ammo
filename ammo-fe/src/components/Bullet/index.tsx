import React, { ReactElement, useState } from 'react';
import makeStyles from '@material-ui/core/styles/makeStyles';
import Box from 'material/Box';
import BulletHeader from 'components/Bullet/BulletHeader';
import BulletDesc from 'components/Bullet/BulletDesc';
import IBullet from 'shared/types/Bullet';
import Palette from 'material/Palette';

const useStyles = makeStyles(() => ({
    root: {
        margin: '2px 2px 0',
        backgroundColor: Palette.BLACK_MED,
    },
    desc: {
        color: Palette.WHITE,
        padding: '7px 16px 24px 16px',
        lineHeight: '24px',
    },
}));

export interface BulletProps {
    bullet: IBullet;
}

const Bullet = ({ bullet }: BulletProps): ReactElement => {
    const classes = useStyles();
    const [collapse, setCollapse] = useState<boolean>(true);

    const toggleCollapse = (): void => setCollapse((prevState) => !prevState);

    return (
        <Box className={classes.root}>
            <BulletHeader
                bullet={bullet}
                collapse={collapse}
                onClick={toggleCollapse}
            />

            <BulletDesc bullet={bullet} collapse={collapse} />
        </Box>
    );
};

export default Bullet;
