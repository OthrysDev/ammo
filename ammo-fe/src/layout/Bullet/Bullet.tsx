import React, { ReactElement, ReactNode } from 'react';
import makeStyles from '@material-ui/core/styles/makeStyles';
import Typography from '@material-ui/core/Typography';
import Grid from '@material-ui/core/Grid';
import Box from '@material-ui/core/Box';
import Bullet from 'layout/Bullet/Bullet';
import IBullet from 'imported/IBullet';

interface BulletProps {
    bullet: IBullet;
}

const useStyles = makeStyles((theme) => ({
    root: {
        backgroundColor: theme.palette.primary.light,
        margin: '3px 7px',
        color: theme.palette.secondary.light,
        padding: '2px 7px',
    },
}));

const MainPannel = ({ bullet }: BulletProps): ReactElement => {
    const classes = useStyles();
    return (
        <Box className={classes.root}>
            <Typography>
                ::{new Date(bullet.date).getMilliseconds()} {bullet.method} |{' '}
                {bullet.url}
            </Typography>
        </Box>
    );
};

export default MainPannel;
