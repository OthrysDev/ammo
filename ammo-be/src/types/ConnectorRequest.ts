import HttpMethod from 'shared/types/HttpMethod';

// Documentation of the HTTP protocol for mandatory / optionnal fields : https://quicwg.org/base-drafts/draft-ietf-quic-http.html#name-http-request-lifecycle

export type ConnectorRequest = {
    url: string;
    method: HttpMethod;
    request: {
        headers: Record<string, string | string[]>;
        body?: unknown;
    };
    response: {
        headers: Record<string, string | string[]>;
        body?: unknown;
        status?: number;
    };
};
