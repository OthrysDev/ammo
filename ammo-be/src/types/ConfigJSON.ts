import HTTPProtocol from 'types/HTTPProtocol';

export default interface ConfigJSON {
    protocol: HTTPProtocol;
    host: string;
    port: number;
    feUrl: string;
}
