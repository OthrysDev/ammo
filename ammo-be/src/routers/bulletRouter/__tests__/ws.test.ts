import app from 'index';
import request from 'supertest';
import { Bullet } from 'shared/types/Bullet';
import { io, Socket } from 'socket.io-client';
import MockDate from 'mockdate';
import { isRecording, ioServer } from 'webSocket';
import {
    connectorRequestMock,
    noMethodConnectorRequestMock,
} from 'routers/bulletRouter/__tests__/mocks/ConnectorRequest.mock';

let socket: Socket;

jest.mock('nanoid', () => ({
    nanoid: (): string => 'AMockedNanoId',
}));

beforeAll((done) => {
    MockDate.set(new Date('2020-10-10 10:00:00'));

    try {
        // Simulate a client connecting to our socket
        socket = io('http://localhost:3001');

        socket.on('connect', done);
    } catch (error) {
        // Keep the console.log for debugging purpose
        console.log('error', error);
        done();
    }
});

afterAll((done) => {
    app.close();

    if (socket.connected) socket.disconnect();

    MockDate.reset();

    done();
});

afterEach(() => {
    socket.off('bullet');
});

describe('[WS] BulletRouter', () => {
    describe('post("/")', () => {
        test('Starting the application - ws should be connected', (done) => {
            expect(socket.connected).toBe(true);

            done();
        });

        test('Sending a connectorRequest - ws should return a bullet', async (done) => {
            socket.on('bullet', ({ bullet }: Record<string, Bullet>) => {
                expect(bullet).toMatchSnapshot();
                done();
            });

            await request(app)
                .post('/')
                .send({ data: connectorRequestMock })
                .expect(200);
        });

        test('Sending an incorrect connectorRequest - ws should not return a bullet', async (done) => {
            socket.on('bullet', () => {
                done.fail(
                    'Should not catch a bullet as the data is incorrect.'
                );
            });

            await request(app)
                .post('/')
                .send({ data: noMethodConnectorRequestMock })
                .expect(400);

            done();
        });
    });

    it('The recording state must be initialized to true', () => {
        expect(isRecording).toBe(true);
    });

    it('Should change the recording state to false', (done) => {
        socket.emit('toggleRecord', (result: boolean) => {
            expect(result).toBe(false);
            expect(isRecording).toBe(false);
            done();
        });
    });

    it('Should not be emitting if the recording state is false', async (done) => {
        const emitSpy = jest.spyOn(ioServer, 'emit');

        await request(app)
            .post('/')
            .send({ data: connectorRequestMock })
            .expect(200);

        expect(emitSpy).toBeCalledTimes(0);
        done();
    });

    it('Should be returning a clear message if the recording state is false', async (done) => {
        await request(app)
            .post('/')
            .send({ data: connectorRequestMock })
            .expect(200)
            .expect((r) => {
                const { message } = r.body;
                expect(message).toMatchSnapshot();

                done();
            });
    });

    it('Should change the recording state to true', (done) => {
        socket.emit('toggleRecord', (result: boolean) => {
            expect(result).toBe(true);
            expect(isRecording).toBe(true);
            done();
        });
    });

    it('Should now be emitting back', async (done) => {
        const emitSpy = jest.spyOn(ioServer, 'emit');

        await request(app)
            .post('/')
            .send({ data: connectorRequestMock })
            .expect(200);

        expect(emitSpy).toBeCalledTimes(1);
        done();
    });
});
