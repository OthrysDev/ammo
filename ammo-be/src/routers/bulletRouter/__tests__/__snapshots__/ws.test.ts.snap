// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[WS] BulletRouter post("/") Sending a connectorRequest - ws should return a bullet 1`] = `
Object {
  "date": "2020-10-10T08:00:00.000Z",
  "id": "AMockedNanoId",
  "method": "POST",
  "request": Object {
    "body": "{ user: {\\"firstName\\":\\"Foo\\",\\"name\\":\\"Bar\\",\\"email\\":\\"foo.bar@mail.com\\",\\"password\\":\\"123456789\\",\\"captcha\\":\\"03AGdBq24FhhL7ma6BDSXC75oSQ3_uKKSYSID3XJyoPQCkwUkkvnZqrEqShLhqus2RMpWWDHq3wXHTXm5qmA9x7mZQup0XYYN9bHj_sFH_i-EmSXfnaH5C3QwImPUd_o-E3f8Zfd6mCIPe98O9_qDHNM-GJ387Fgajj5PbFZZuj4QG2iTPeEM1IoaQdO3Zn-XmMYqU3YU18O3rMMfpLiVtk7i2k8zLPpFHiDp3US3Tc91fr0ot2xUYOGTPwM58LgnWrX31X-oezUEk8EagkPD2IAwgh0vu55DHE7HOSaARFatfeoUgNwgVTnsRmfhN1gGnxGIUKG8iklaVEYG34Vg4CZ800uxQIUALLGdbNnFV5Pwxkf4XiaOiKY-XUXoTBNu_THbbNES08Itj2N_Qj7BhTG8mvVV3sjVsS0iVHkGxnssuRZaFbZ6JciD6QJG8cB1fksMYf5_CBCil\\"}}",
    "headers": Object {
      "accept": "application/json, text/plain, */*",
      "accept-encoding": "gzip, deflate, br",
      "accept-language": "fr-FR,fr;q=0.9,en-US;q=0.8,en;q=0.7",
      "authorization": "Bearer c9da47d8f9bf16af86e325adc1e191ffa6f535f3",
      "client-uuid": "3T3iVH_DKKSHlbD9dETTvZTqUApkxNKMdlY-b0z",
      "connection": "keep-alive",
      "content-length": "417873",
      "content-type": "multipart/form-data; boundary=----WebKitFormBoundarylgQY9Tli7bNsE3LG",
      "host": "localhost:3000",
      "origin": "http://localhost:3001",
      "referer": "http://localhost:3001/",
      "sec-fetch-dest": "empty",
      "sec-fetch-mode": "cors",
      "sec-fetch-site": "same-site",
      "user-agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/86.0.4240.198 Safari/537.36",
    },
  },
  "response": Object {
    "body": "{\\"success\\":true,\\"message\\":{\\"user\\":{\\"_id\\":\\"5fb3a5ae5be2f60fc786d90b\\",\\"firstName\\":\\"Foo\\",\\"name\\":\\"Bar\\",\\"email\\":\\"foo.bar@mail.com\\"}}}",
    "headers": Object {
      "access-control-allow-credentials": "true",
      "access-control-allow-origin": "http://localhost:3001",
      "access-control-expose-headers": "jwtToken",
      "content-length": "229",
      "content-type": "application/json; charset=utf-8",
      "etag": "W/\\"e5-L2Zax/+Z+lIbnuvKRq1seHacyG8\\"",
      "vary": "Origin",
      "x-powered-by": "Express",
    },
    "status": 200,
  },
  "url": "http://localhost:3000/user",
}
`;

exports[`[WS] BulletRouter Should be returning a clear message if the recording state is false 1`] = `"The recorder has been paused, we will not process your request any further"`;
